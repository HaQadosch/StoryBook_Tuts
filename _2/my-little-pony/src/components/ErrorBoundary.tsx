import React, { ErrorInfo } from 'react';

export class ErrorBoundary extends React.Component<{}, { hasError: boolean }> {
  constructor(props: React.ReactChildren) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: Error) {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error: Error, info: ErrorInfo) {
    // You can also log the error to an error reporting service
    console.error(error, info);
  }

  render() {
    if (this.state.hasError) {
      // You can render any custom fallback UI
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children;
  }
}
