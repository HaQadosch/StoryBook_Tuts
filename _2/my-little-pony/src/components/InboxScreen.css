.page.lists-show {
  max-width: 400px;
  width: 100%;
}
