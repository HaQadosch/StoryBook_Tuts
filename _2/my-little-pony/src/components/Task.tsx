import React from 'react';

export enum TaskState {
  INBOX,
  PINNED,
  ARCHIVED,
}

export type TTask = {
  id: string;
  title?: string;
  state?: TaskState;
  updatedAt?: Date;
};

interface ITask {
  task: TTask;
  onArchive?: React.ChangeEventHandler<HTMLSelectElement>;
  onPin?: React.ChangeEventHandler<HTMLSelectElement>;
}

export const Task: React.FC<ITask> = ({ task: { id, title } }) => {
  return (
    <div className='list-item'>
      <input type='text' value={title} name={title} id={id} aria-label={title} />
    </div>
  );
};
