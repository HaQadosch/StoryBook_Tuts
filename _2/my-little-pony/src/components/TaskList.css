.list-items {
  display: flex;
  flex-direction: column;
}

.PINNED {
  order: 1;
}

.INBOX {
  order: 2;
}
