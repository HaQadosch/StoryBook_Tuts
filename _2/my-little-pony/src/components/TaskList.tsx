import React from 'react';
import { Task, TTask, TaskState } from './Task';

interface ITaksList {
  loading: boolean;
  tasks: TTask[];
  onArchive(id: string): void;
  onPin(id: string): void;
}

export const TaskList: React.FC<ITaksList> = ({ loading, tasks, onArchive, onPin }) => {
  const events = { onArchive, onPin };

  const loadingRow = (
    <div className='loading-item'>
      <span className='glow-checkbox' />
      <span className='glow-text'>
        <span>Loading</span> <span>cool</span> <span>state</span>
      </span>
    </div>
  );

  if (loading) {
    return (
      <div className='list-items'>
        {loadingRow}
        {loadingRow}
        {loadingRow}
        {loadingRow}
        {loadingRow}
        {loadingRow}
      </div>
    );
  }

  if (tasks.length === 0) {
    return (
      <div className='list-items'>
        <div className='wrapper-message'>
          <span className='icon-check' />
          <div className='title-messag'>You have no tasks.</div>
          <div className='subtitle-message'>Sit back and relax</div>
        </div>
      </div>
    );
  }

  // We want the pinned tasks on top.
  const splitFilter = (all: any[]) => (condition: (test: any) => boolean) =>
    all.reduce(([yes, no], cur) => (condition(cur) ? [[...yes, cur], no] : [yes, [...no, cur]]), [[], []]);

  const [pinned, notPinned] = splitFilter(tasks)(({ state }) => state === TaskState.PINNED);

  const tasksInOrder = [...pinned, ...notPinned];

  return (
    <div className='list-items'>
      {tasksInOrder.map(task => (
        <Task key={task.id} task={task} {...events} />
      ))}
    </div>
  );
};
