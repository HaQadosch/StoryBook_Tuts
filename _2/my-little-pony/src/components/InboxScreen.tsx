import React from 'react';
import { TaskList } from './TaskList';
import './InboxScreen.css';
import { defaultTasks } from '../utils/genTasks';

export const InboxScreen: React.FC<{ error?: Error; errorInfo?: React.ErrorInfo }> = ({ error, errorInfo }) => {
  const errorComponent = (
    <div className='wrapper-message'>
      <span className='icon-face-sad' />
      <div className='title-message'>Oh no!</div>
      <div className='subtitle-message'>Somethig went wrong</div>
    </div>
  );

  const okComponent = (
    <>
      <nav>
        <h1 className='title-page'>
          <span className='title-wrapper'>TaskBox</span>
        </h1>
      </nav>
      <TaskList loading={false} tasks={defaultTasks} onArchive={() => {}} onPin={() => {}} />
    </>
  );

  return <div className='page lists-show'>{error ? errorComponent : okComponent}</div>;
};
