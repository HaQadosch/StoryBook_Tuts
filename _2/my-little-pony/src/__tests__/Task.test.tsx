import React from 'react';
import ReactDOM from 'react-dom';
import { render } from 'react-testing-library';
import produce from 'immer';

import { Task, TTask } from '../components/Task';

const baseTask: TTask = {
  id: 'testId',
  title: 'testTitle',
  state: 'INBOX',
};

// Insert the creation date as updatedAt.
const createTask = (base: TTask): TTask =>
  produce(base, draft => {
    draft.updatedAt = new Date();
  });

it('renders without crashing', () => {
  const task = createTask(baseTask);
  const div = document.createElement('div');
  ReactDOM.render(<Task task={task} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('Task does its thiny', () => {
  const task = createTask(baseTask);
  const { debug, asFragment } = render(<Task task={task} />);
  expect(asFragment()).toMatchSnapshot();
  debug(asFragment());
});
