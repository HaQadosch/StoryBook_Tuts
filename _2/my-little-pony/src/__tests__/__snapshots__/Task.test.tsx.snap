// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task does its thiny 1`] = `
<DocumentFragment>
  <div
    class="list-item"
  >
    <input
      aria-label="testTitle"
      id="testId"
      name="testTitle"
      type="text"
      value="testTitle"
    />
  </div>
</DocumentFragment>
`;
